### ISSUE TITLE
[BUG] tensorflow hang when use sparse_operation_kit with multiple embedding layers, maybe caused by gradient tape.

### ISSUE THREAD

**Issue by marsmiao** [NVIDIA-Merlin/HugeCTR#260]
**Describe the bug**

When use multiple embedding layers (cause we have multiple embedding_vec_size), and there maybe a shared variable usage because we want to distributed and all2all dense to use the same table. In training loop, I carefully use tf.control_dependencies to make nccl launched in order, while tf.gradient_tape maybe parallel launch nccl (embedding layers backward propagation), there will be hanged in V100 or A10. However, two-T4 will not be hanged on the same model.

**To Reproduce**
I can give a code snippet in my github later.

**Expected behavior**
For tensorflow2 that must be used with gradient tape, how can we fix this?

**Screenshots**
![image](https://user-images.githubusercontent.com/56477651/140863008-8308426b-dd13-4fb5-a123-5184ef055a5c.png)

**Environment (please complete the following information):**
 - Docker image: tensorflow-merlin-0.6

**Comment by marsmiao** [comment]
@Jianbing-D sry to bother you again, this is the problem that I'm desperate to find solution to.

**Comment by Jianbing-D** [comment]
@marsmiao Please provide the aforementioned code snippet.

**Comment by marsmiao** [comment]
> @marsmiao Please provide the aforementioned code snippet.

https://github.com/marsmiao/sok-demo/tree/master please have a look, thank you.

**Comment by marsmiao** [comment]
> @marsmiao Please provide the aforementioned code snippet.

@Jianbing-D forgot to add access, it is done now.

**Comment by Jianbing-D** [comment]
There are several differences between your usage and SOK's design.
1. https://github.com/marsmiao/sok-demo/blob/master/main.py#L107 There is no `prefetch` argument in `sok.DistributedEmbedding`. Could you please elaborate what the `prefetch` here means, so that we can tell whether it is related to the hanging.
2. https://github.com/marsmiao/sok-demo/blob/master/main.py#L116 SOK will create EmbeddingVariable per each EmbeddingLayer. Although you have modified the `all2all_dense_embedding.py` to make that embedding layer takes already created Variable, in the backend, each EmbeddingVariable can only be bound to one Embedding Layer. So that after  this line, `self.embedding_layer_25.embedding_variable` only belongs to `self.embedding_layer_dense_25`. 
3. Unfortunately, SOK does not support sharing Variable among different embedding layers. If you really need such feature, please file an issue and describe what you want. Then we will try to make it happen.

**Comment by marsmiao** [comment]
> There are several differences between your usage and SOK's design.
> 
> 1. https://github.com/marsmiao/sok-demo/blob/master/main.py#L107 There is no `prefetch` argument in `sok.DistributedEmbedding`. Could you please elaborate what the `prefetch` here means, so that we can tell whether it is related to the hanging.
> 2. https://github.com/marsmiao/sok-demo/blob/master/main.py#L116 SOK will create EmbeddingVariable per each EmbeddingLayer. Although you have modified the `all2all_dense_embedding.py` to make that embedding layer takes already created Variable, in the backend, each EmbeddingVariable can only be bound to one Embedding Layer. So that after  this line, `self.embedding_layer_25.embedding_variable` only belongs to `self.embedding_layer_dense_25`.
> 3. Unfortunately, SOK does not support sharing Variable among different embedding layers. If you really need such feature, please file an issue and describe what you want. Then we will try to make it happen.
> 4. For sharing Variable among different embedding layers, there exist a work around for you:
>    ![image](https://user-images.githubusercontent.com/69858819/141036179-cc438279-ca16-4e1a-a3aa-410a17ff9fd8.png)
> 
> To make your codes clean, you can add another python wrapper of `sok.All2AllDenseEmbedding`, and make those wrapper instances share the same `sok.All2AllDenseEmbedding`.

1. Sorry, this 'prefetch' is what we did to use mos-like feature and it's still in progress. sorry to mislead you and this option should be ignored.
2. Yes, you are right and this is a hackable solution, but for a shared-table which many of them is sparse feature, use the 4 you mentioned will be a decrease in performance. Although it maybe not a solid solution, but we looked at the code inside sparse operation kit, the lookup will still be operated on the same variable and the backward is still normal to what we expected, so maybe you can consider this is a workaround.
3. what we find for now, the concat in multiple embedding vectors maybe caused hang because backward will not be what we expected one by one. however when we increase batch-size, even tf.control_dependencies won't be able to solve the hang issue, so maybe you have some ideas on this issue?

Thanks a lot for your response. Wish you a good day.

**Comment by Jianbing-D** [comment]
Perhaps you can try `tf.distribute.MultiWorkerMirroredStrategy` or `Horovod`, where single CPU process & single CPU thread are bounded for single GPU.

**Comment by zehuanw** [comment]
Hi @marsmiao , thank you for your feedback! Does @Jianbing-D 's solution help you? We are looking forward your feedbacks.

**Comment by marsmiao** [comment]
> Hi @marsmiao , thank you for your feedback! Does @Jianbing-D 's solution help you? We are looking forward your feedbacks.

sorry for the late reply. As far as I'm concerned, precisely control the concatenate order of sparse operation kit's embedding layer and it will be solved.
